### PR TITLE
Refactor VFIO code

### PR DESCRIPTION
**Pull request by n-eiling**

### CODE REVIEW

File: include/villas/kernel/vfio_container.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
+/** Virtual Function IO wrapper around kernel API
+ *
+ * @file
+ * @author Niklas Eiling <niklas.eiling@eonerc.rwth-aachen.de>
+ * @author Steffen Vogel <svogel2@eonerc.rwth-aachen.de>
+ * @author Daniel Krebs <github@daniel-krebs.net>
+ * @copyright 2022, Niklas Eiling
+ * @copyright 2014-2021, Steffen Vogel
+ * @copyright 2018, Daniel Krebs
```
Comment (stv0g) [inline include/villas/kernel/vfio_container.hpp:9]: Please add licence to header. Use one of the other files as a templates as this will simplify project-wide search & replace lateron.

File: include/villas/kernel/vfio_container.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+#pragma once
+
+#include <list>
+#include <vector>
+#include <memory>
+
+#include <linux/vfio.h>
+#include <sys/mman.h>
+
+#include "vfio_group.hpp"
```
Comment (stv0g) [inline include/villas/kernel/vfio_container.hpp:21]: please use `#include <villas/...>`

File: include/villas/kernel/vfio_device.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+#include <list>
+#include <vector>
+#include <memory>
+#include <string>
+
+#include <linux/vfio.h>
+#include <sys/mman.h>
+
+#include "pci.hpp"
```
Comment (stv0g) [inline include/villas/kernel/vfio_device.hpp:22]: Fix include style

File: include/villas/graph/directed.hpp [hunk 4]
```diff
@@ -42,7 +42,7 @@ class DirectedGraph {
 
 	std::shared_ptr<VertexType> getVertex(VertexIdentifier vertexId) const
 	{
-		if (vertexId >= lastVertexId)
+		if (vertexId >= vertices.size())
```
Comment (stv0g) [inline include/villas/graph/directed.hpp:45]: Why is this change required?

File: include/villas/kernel/vfio_device.hpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	std::vector<struct vfio_irq_info> irqs;
+	std::vector<struct vfio_region_info> regions;
+	std::vector<void*> mappings;
+
+	/**< libpci handle of the device */
+	const kernel::pci::Device *pci_device;
+
+	Logger log;
+};
+} /* namespace vfio */
```
Comment (stv0g) [inline include/villas/kernel/vfio_device.hpp:96]: Missing new line

File: include/villas/kernel/vfio_group.hpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+#pragma once
+
+#include <list>
+#include <vector>
+#include <memory>
+
+#include <linux/vfio.h>
+#include <sys/mman.h>
+
+#include "vfio_device.hpp"
```
Comment (stv0g) [inline include/villas/kernel/vfio_group.hpp:21]: Fix include style

File: lib/kernel/vfio.cpp [hunk 7]
```diff
@@ -59,11 +59,27 @@ static const char *vfio_pci_irq_names[] = {
 
 namespace villas {
 
+void panicResetEverything()
+{
+	int ret = 0;
+	for (auto& group : container->getGroups()) {
+		for (auto& device : group->getDevices()) {
+			ret += device->reset();
+
```
Comment (stv0g) [inline lib/kernel/vfio.cpp:68]: Superfluous empty line

File: lib/kernel/vfio.cpp [hunk 8]
```diff
@@ -59,11 +59,27 @@ static const char *vfio_pci_irq_names[] = {
...
+	}
+	if (ret > 0)
+		logging.get("kernel:vfio")->error("panic reset return errors.");
+}
 
 Container::Container()
 	: iova_next(0)
 {
 	Logger logger = logging.get("kernel:vfio");
+	spdlog::set_level(spdlog::level::debug);
```
Comment (stv0g) [inline lib/kernel/vfio.cpp:82]: This set the log level globally?

Isnt there a `logger->set_level()`?

File: lib/kernel/vfio.cpp [hunk 9]
```diff
@@ -735,6 +758,7 @@ Group::~Group()
 	if (fd < 0)
 		logger->debug("Destructing group that has not been attached");
 	else {
+		logger->debug("unsetting group container");
```
Comment (stv0g) [inline lib/kernel/vfio.cpp:761]: Please start log messages with capital letter.

File: lib/kernel/vfio.cpp [hunk 10]
```diff
@@ -107,6 +123,11 @@ Container::Container()
 	logger->debug("Version:    {:#x}", version);
 	logger->debug("Extensions: {:#x}", extensions);
 	logger->debug("IOMMU:      {}", hasIommu ? "yes" : "no");
+
```
Comment (stv0g) [inline lib/kernel/vfio.cpp:126]: Why is this not done in the destructor?

File: lib/kernel/vfio_device.cpp [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+
+#include <algorithm>
+#include <string>
+#include <sstream>
+#include <limits>
+
+#include <cstdlib>
+#include <cstring>
+
+#include <cstdint>
```
Comment (stv0g) [inline lib/kernel/vfio_device.cpp:27]: I dont think all these are needed here..

**Review by stv0g — CHANGES_REQUESTED** [review]

File: include/villas/kernel/vfio_container.hpp (hunk 1 above)
Comment (n-eiling) [reply include/villas/kernel/vfio_container.hpp:9]: I did not find any file with a license.
